zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1088 @@
+        // scanned range.
+        subtree_bounds
+            .map(|(min_idx, max_idx)| {
+                let range_min = if *min_idx > 0 {
+                    // get the block height of the end of the previous shard
+                    sapling_shard_end(*min_idx - 1)?
+                } else {
+                    // our lower bound is going to be the Sapling activation height
+                    params.activation_height(NetworkUpgrade::Sapling)
+                };
+
+                // get the block height for the end of the current shard
+                let range_max = sapling_shard_end(*max_idx)?;
+
+                Ok::<Range<BlockHeight>, rusqlite::Error>(Range {
+                    start: range.start.min(range_min.unwrap_or(range.start)),
+                    end: range.end.max(range_max.unwrap_or(range.end)),
+                })
+            })
+            .transpose()
+            .map_err(SqliteClientError::from)
+    }?;
+
+    let query_range = extended_range.clone().unwrap_or_else(|| range.clone());
+
+    let scanned = ScanRange::from_parts(range.clone(), ScanPriority::Scanned);
+    let extensions = if let Some(extended) = extended_range {
+        vec![
+            ScanRange::from_parts(extended.start..range.start, ScanPriority::FoundNote),
+            ScanRange::from_parts(range.end..extended.end, ScanPriority::FoundNote),
+        ]
+    } else {
+        vec![]
+    };
+
+    replace_queue_entries(
+        conn,
+        &query_range,
+        Some(scanned).into_iter().chain(extensions.into_iter()),
+    )?;
+
+    Ok(())
+}
+
+pub(crate) fn update_chain_tip<P: consensus::Parameters>(
+    conn: &rusqlite::Transaction<'_>,
+    params: &P,
+    new_tip: BlockHeight,
+) -> Result<(), SqliteClientError> {
+    // `ScanRange` uses an exclusive upper bound.
+    let chain_end = new_tip + 1;
+
+    // Read the maximum height from the shards table.
+    let shard_start_height = conn.query_row(
+        "SELECT MAX(subtree_end_height)
+        FROM sapling_tree_shards",
+        [],
+        |row| Ok(row.get::<_, Option<u32>>(0)?.map(BlockHeight::from)),
+    )?;
+
+    // Create a scanning range for the fragment of the last shard leading up to new tip.
+    let shard_entry = shard_start_height
+        .filter(|h| h < &chain_end)
+        .map(|h| ScanRange::from_parts(h..chain_end, ScanPriority::ChainTip));
+
+    // Create scanning ranges to either validate potentially invalid blocks at the wallet's view
+    // of the chain tip, or connect the prior tip to the new tip.
+    let tip_entry = block_height_extrema(conn)?.map(|(_, prior_tip)| {
+        // If we don't have shard metadata, this means we're doing linear scanning, so create a
+        // scan range from the prior tip to the current tip with `Historic` priority.
+        if shard_entry.is_none() {
+            ScanRange::from_parts(prior_tip..chain_end, ScanPriority::Historic)
```

Comment:
How do we know that `prior_tip <= chain_end`? If not then `ScanRange::from_parts` will panic.

Reply:
I think @nuttycom was assuming that `update_chain_tip` would not be called frequently (and thus be vulnerable to the chain tip going backwards, because if you're calling this less frequently than on the order of minutes, you won't have this problem). But you're correct that if a reorg occurs that causes the chain height to decrease (which is technically possible as chain work can still increase) then there's a panic opportunity here.

The only information we have available to us here is the tip height, so I think all we can do here is shrug and do nothing (as far as `tip_entry` goes), and wait for the next time the caller calls `update_tip` with a tip higher than the latest the wallet DB is aware of.

Reply:
Ok. That requires explicitly checking for that case, no?

Reply:
Pushed a commit that does nothing in this scenario.